src/app/clusters/device-energy-management-server/device-energy-management-server.h

```diff
@@ -0,0 +1,218 @@
+     *   reflect its current state.
+     *
+     * @return  Success if the ESA is resumed, otherwise returns other IM_Status.
+     */
+    virtual Status ResumeRequest() = 0;
+
+    /**
+     * @brief Delegate handler for ModifyForecastRequest
+     *
+     *   If the ESA supports FA, and the ESAState is not UserOptOut it SHALL attempt to adjust its power forecast.
+     *   This allows a one or more modifications in a single command by sending a list of modifications (one for each 'slot').
+     *   Attempts to modify slots which have already past, SHALL result in the entire command being rejected.
+     *   If the ESA accepts the requested Forecast then it SHALL update its Forecast attribute (incrementing its ForecastId)
+     *   and run the revised Forecast as its new intended operation.
+     *
+     * @param forecastId Indicates the ESA ForecastId that is to be modified.
+     * @param slotAdjustments List of adjustments to be applied to the ESA, corresponding to the expected ESA forecastId.
+     * @return  Success if the entire list of SlotAdjustmentStruct are accepted, otherwise the command
+     *          SHALL be rejected returning other IM_Status.
+     */
+    virtual Status ModifyForecastRequest(const uint32_t forecastId,
+                                         const DataModel::DecodableList<Structs::SlotAdjustmentStruct::Type> & slotAdjustments) = 0;
+
+    /**
+     * @brief Delegate handler for RequestConstraintBasedForecast
+     *
+     *   The ESA SHALL inspect the requested power limits to ensure that there are no overlapping elements. The ESA
+     *   manufacturer may also reject the request if it could cause the user’s preferences to be breached (e.g. may
+     *   cause the home to be too hot or too cold, or a battery to be insufficiently charged).
+     *   If the ESA can meet the requested power limits, it SHALL regenerate a new Power Forecast with a new ForecastId.
+     *
+     * @param constraints  Sequence of turn up/down power requests that the ESA is being asked to constrain its operation within.
+     * @return  Success if successful, otherwise the command SHALL be rejected returning other IM_Status.
+     */
+    virtual Status
+    RequestConstraintBasedForecast(const DataModel::DecodableList<Structs::ConstraintsStruct::Type> & constraints) = 0;
+
+    // ------------------------------------------------------------------
+    // Get attribute methods
+    virtual ESATypeEnum GetESAType()                                                 = 0;
+    virtual bool GetESACanGenerate()                                                 = 0;
+    virtual ESAStateEnum GetESAState()                                               = 0;
+    virtual int64_t GetAbsMinPower()                                                 = 0;
+    virtual int64_t GetAbsMaxPower()                                                 = 0;
+    virtual PowerAdjustmentCapability::TypeInfo::Type GetPowerAdjustmentCapability() = 0;
```

Comment:
This API shape requires the delegate to have the memory backing this list live forever, since it can't be sure when we are done with it. I doubt that's intended.

Reply:
Actually that was the intention - at least in my mind the application would use the delegate to hold these relatively complex structures. From the ESA application side to can set the Forecast into the delegate, and then when a Matter client / controller makes a read request or subscribes to an attribute, then it gets the latest view.

Equally when the application wants to follow a schedule then it can use the forecast as its master copy.
[This comment applies to the PowerAdjustmentCapability, but the intention here is the same as for the forecast].

I'm trying to picture an alternative implementation where the application also has its own copy - there's nothing to stop that from happening, but then it's wasteful of memory having two.

In my earlier experience of developing with Matter, we were short on memory, and needed to comply to the matter types, so we decided just to use the ember framework to store our variables, which our application code just did Get / Set calls.

Clearly there are other models where this could be 'freed', but since reads happen asynchronously the then when the delegate is asked by this file for the PowerAdjustmentCapability then it could construct it on the fly (be having calls into the user application code) or just return the latest copy stored in the delegate.

I'm not sure how I would change this at the device-energy-management-server.cpp layer though? It just calls mDelegate.GetPowerAdjustmentCapability().

The delegate in the examples folder is just an example, but I would expect more manufacturers to simply adopt it and customise it a bit if they needed to.

Reply:
Ok I now understand the issue with lists being chunk encoded over time.

I've raised a follow up issue to track and fix.
https://github.com/project-chip/connectedhomeip/issues/31336